docs/docs-using-extending.md

```diff
@@ -1,28 +1,94 @@
+[lib]
+name = "my_plugin_name"
+crate-type = ["cdylib"]
+
+[dependencies]
+javy-plugin-api = "1.0.0"
+```
+
+And `src/lib.rs` should look like:
+
+```rust
+use javy_plugin_api::import_namespace;
+use javy_plugin_api::javy::quickjs::prelude::Func;
+use javy_plugin_api::javy::Config;
+
+import_namespace!("my_plugin_name");
+
+#[export_name = "initialize_runtime"]
+pub extern "C" fn initialize_runtime() {
+    let mut config = Config::default();
+    config
+        .text_encoding(true)
+        .javy_stream_io(true);
+
+    javy_plugin_api::initialize_runtime(config, |runtime| {
+        runtime
+            .context()
+            .with(|ctx| {
+                ctx.globals()
+                    .set("isThisAPlugin", Func::from(|| "yes it is"))
+            })
+            .unwrap();
+        runtime
+    })
+    .unwrap();
+}
+```
+
+You can then run `cargo build --target=wasm32-wasip1 --release` to create a
+Wasm module. Then you need to run
+
+```
+javy init-plugin <path_to_plugin> -o <path_to_initialized_module>`
+```
+
+which will validate and initialize the Javy runtime. This `javy init-plugin`
+step is required for the plugin to be useable by the Javy CLI.
 
 See our documentation on [using complex data types in Wasm
 functions](./contributing-complex-data-types.md) for how to support Wasm
 functions that need to use byte arrays, strings, or structured data.
 
-For a visual representation of how we expect forks to consume our crates:
+## The full plugin API
 
-```mermaid
-flowchart TD
-  your-cli --> wasm
-  subgraph wasm[your_plugin.wasm]
-  your-plugin --> javy[upstream javy]
-  javy --> rquickjs
-  end
-```
+The Wasm API the Javy CLI expects Javy plugins to expose is the following:
+- Exported `initialize_runtime() -> ()` function. This will initialize a mutable
+  global containing the Javy runtime for use by `compile_src` and `invoke`.
+- Exported
+  `canonical_abi_realloc(orig_ptr: i32, orig_len: i32, new_ptr: i32, new_len: i32) -> ptr: i32`
+  function. This is used to allocate memory in the plugin module.
+- Exported `compile_src(src_ptr: i32, src_len: i32) -> bytecode_wide_ptr: i32`
+  function. This is used to compile JavaScript source code to QuickJS bytecode.
+  The return pointer points to a tuple of
+  `(bytecode_ptr: i32, bytecode_len: i32)` in the plugin instance's linear
+  memory.
+- Exported
+  `invoke(bytecode_ptr: i32, bytecode_len: i32, fn_name_ptr: i32, fn_name_len: i32) -> ()`
+  function. This is used to evaluate the JavaScript code and optionally to
+  call an exported JS function if `fn_name_ptr` is not `0`.
+- Custom section named `import_namespace` containing a UTF-8 encoded string.
+  This is used to determine the namespace that will be used for the Wasm
+  imports in dynamically linked modules built with this plugin.
+
+The `javy-plugin-api` crate will export implementations of all required exported
+functions except `initialize_runtime`. `import_namespace!` will define the
+`import_namespace` custom section.
```

Comment:
Minor nit: what do you think of putting this information inside a table to improve readability? It'll for sure be less readable in markdown format, but potentially more readable when rendered.